zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementTreePathBuilder.java

```diff
@@ -0,0 +1,86 @@
+import com.google.common.hash.Hashing;
+import io.camunda.zeebe.engine.state.immutable.ElementInstanceState;
+import io.camunda.zeebe.engine.state.instance.ElementInstance;
+import java.nio.charset.Charset;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+public class ElementTreePathBuilder {
+  private ElementInstanceState elementInstanceState;
+  private Long elementInstanceKey;
+  private ElementTreePathProperties properties;
+  private final HashFunction hashFunction;
+
+  public ElementTreePathBuilder() {
+    hashFunction = Hashing.murmur3_128();
+  }
+
+  public ElementTreePathBuilder withElementInstanceState(
+      final ElementInstanceState elementInstanceState) {
+    this.elementInstanceState = elementInstanceState;
+    return this;
+  }
+
+  public ElementTreePathBuilder withElementInstanceKey(final long elementInstanceKey) {
+    this.elementInstanceKey = elementInstanceKey;
+    return this;
+  }
+
+  public ElementTreePathProperties build() {
+    Objects.requireNonNull(elementInstanceState, "elementInstanceState cannot be null");
+    Objects.requireNonNull(elementInstanceKey, "elementInstanceKey cannot be null");
+    properties =
+        new ElementTreePathProperties(new LinkedList<>(), new LinkedList<>(), new LinkedList<>());
+    buildElementTreePathProperties(elementInstanceKey);
+    return properties;
+  }
+
+  private void buildElementTreePathProperties(final long elementInstanceKey) {
+    final List<Long> elementInstancePath = new LinkedList<>();
+    elementInstancePath.add(elementInstanceKey);
+    ElementInstance instance = elementInstanceState.getInstance(elementInstanceKey);
+    long parentElementInstanceKey = instance.getParentKey();
+    while (parentElementInstanceKey != -1) {
+      instance = elementInstanceState.getInstance(parentElementInstanceKey);
+      elementInstancePath.addFirst(parentElementInstanceKey);
+      parentElementInstanceKey = instance.getParentKey();
+    }
+    properties.elementInstancePath.addFirst(elementInstancePath);
+    final var processInstanceRecord = instance.getValue();
+    properties.processDefinitionPath.addFirst(processInstanceRecord.getProcessDefinitionKey());
+
+    final long callingElementInstanceKey = processInstanceRecord.getParentElementInstanceKey();
+    if (callingElementInstanceKey != -1) {
+      properties.callingElementPath.addFirst(getCallActivityIdHash(callingElementInstanceKey));
+      buildElementTreePathProperties(callingElementInstanceKey);
+    }
+  }
+
+  private Long getCallActivityIdHash(final long callingElementInstanceKey) {
+    final ElementInstance callActivityElementInstance =
+        elementInstanceState.getInstance(callingElementInstanceKey);
+    final var callActivityInstanceRecord = callActivityElementInstance.getValue();
+
+    final HashCode hashCode =
+        hashFunction.hashString(
+            callActivityInstanceRecord.getElementId(), Charset.defaultCharset());
+    return hashCode.asLong();
```

Comment:
That's a good pint. 
Originally, I was looking to use [goodFastHash](https://guava.dev/releases/snapshot/api/docs/com/google/common/hash/Hashing.html#goodFastHash(int)) with 64 bits, to provide a balance between performance and collision safety. But I saw in the implementation it uses the `Murmur3_128HashFunction` for anything above 32 bits anyway.

I could have used `toString()` as well. And I could still change it use strings if we see a need.

Reply:
switched to `HashCode.toString()` which converts the hash to hexadecimal string, using the full 128 bits to ensure less likelihood for collision.

Reply:
💭 Just a side note:
I've asked this [here](https://camunda.slack.com/archives/C077SHZS6R0/p1719924818601759) before, and I don't think we know the answer right now.
The hexadecimal string representing `128 bit` hash will be of `32 char` length. If we (have a way to) know that most CallActivity Ids are shorter than this length, then hashing this way is working against us not for us. If it's the other way around, then that's okay.
In the case of using `long`, size wouldn't have mattered, because long is almost always more memory efficient. yet, compromising on some of the anti-collision capabilities.

The above is a trade-off I can't make my mind about without data to support, but we unfortunately, we don't have this data AFAIK.

Reply:
I think one important property of hashing (or any other form of encoding) is that it protects against pathological cases where ids are _extremely_ long. So that's something we need to consider too, not just the average case.

> The hexadecimal string representing 128 bit hash will be of 32 char length
We could also encode it as two longs instead. Or use base64 which only take 22 bytes.

In general I agree though, using a 128 bit hash is probably not very efficient for the common case.